### ISSUE TITLE
Bindparam error with Tuple

### ISSUE THREAD

**Issue by tvuotila** [sqlalchemy/sqlalchemy#5654]
**Describe the bug**

Higher level problem is that [selectinload](https://docs.sqlalchemy.org/en/13/orm/loading_relationships.html#sqlalchemy.orm.selectinload) doesn't process each primary key with the correct bind value processor.

I believe that I have narrowed the problem down to "Expanding [Bindparam](https://docs.sqlalchemy.org/en/13/core/sqlelement.html#sqlalchemy.sql.expression.bindparam) does the bind value conversion wrongly if using [Tuple](https://docs.sqlalchemy.org/en/13/core/sqlelement.html#sqlalchemy.sql.expression.Tuple)".

The problem is that the same bind processor is used for all clauses of the tuple.

**Expected behavior**
Each clause is processed with the bind processor of the clause type.

**To Reproduce**

```py
from datetime import date

from sqlalchemy import Column, create_engine, Integer
from sqlalchemy.dialects.sqlite import DATE
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import relationship, selectinload, sessionmaker
from sqlalchemy.schema import ForeignKeyConstraint
from sqlalchemy.sql.expression import bindparam, tuple_

Base = declarative_base()


class User(Base):
    __tablename__ = 'user'
    batch_date = Column(DATE, primary_key=True)
    id = Column(Integer, primary_key=True)


class Address(Base):
    __tablename__ = 'address'
    id = Column(Integer, primary_key=True)
    batch_date = Column(DATE, nullable=False)
    user_id = Column(Integer, nullable=False)
    user = relationship(
        'User',
        foreign_keys=[batch_date, user_id],
        backref='addresses',
    )
    __table_args__ = (
        ForeignKeyConstraint(
            [batch_date, user_id],
            ['user.batch_date', 'user.id'],
            ondelete='CASCADE'
        ),
    )


engine = create_engine('sqlite:///:memory:', echo=True)
Base.metadata.create_all(engine)
Session = sessionmaker(bind=engine)
session = Session()

# Insert some test data
timestamp = date(2020, 10, 18)
user = User(id=1, batch_date=timestamp)
session.add(Address(id=1, batch_date=timestamp, user=user))
session.commit()

# My original error was with selectinload
session.query(Address).options(selectinload(Address.user)).all()

# This works
(
    session.query(User)
    .filter(
        tuple_(User.batch_date, User.id)
        .in_([(date(2020, 10, 18), 1)])
    ).all()
)

# The following bindparam version fails.
# It is similar construct to the one the selectinload uses.
(
    session.query(User)
    .filter(
        tuple_(User.batch_date, User.id)
        .in_(bindparam('primary_keys', expanding=True))
    ).params(primary_keys=[(date(2020, 10, 18), 1)])
    .all()
)
```

**Error**

For the selectinload query:
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
~/.virtualenvs/incy/lib/python3.8/site-packages/sqlalchemy/engine/base.py in _execute_context(self, dialect, constructor, statement, parameters, *args)
   1203
-> 1204             context = constructor(dialect, self, conn, *args)
   1205         except BaseException as e:

~/.virtualenvs/incy/lib/python3.8/site-packages/sqlalchemy/engine/default.py in _init_compiled(cls, dialect, connection, dbapi_connection, compiled, parameters)
    838                     if key in processors:
--> 839                         param.append(processors[key](compiled_params[key]))
    840                     else:

~/.virtualenvs/incy/lib/python3.8/site-packages/sqlalchemy/dialects/sqlite/base.py in process(value)
    839             else:
--> 840                 raise TypeError(
    841                     "SQLite Date type only accepts Python "

TypeError: SQLite Date type only accepts Python date objects as input.

The above exception was the direct cause of the following exception:

StatementError                            Traceback (most recent call last)
<ipython-input-1-5a25fdc9cadd> in <module>
     48
     49 # My original error was with selectinload
---> 50 session.query(Address).options(selectinload(Address.user)).all()
     51
     52 # This works

~/.virtualenvs/incy/lib/python3.8/site-packages/sqlalchemy/orm/query.py in all(self)
   3344                 :ref:`faq_query_deduplicating`
   3345         """
-> 3346         return list(self)
   3347
   3348     @_generative(_no_clauseelement_condition)

~/.virtualenvs/incy/lib/python3.8/site-packages/sqlalchemy/orm/loading.py in instances(query, cursor, context)
     98     except Exception:
     99         with util.safe_reraise():
--> 100             cursor.close()
    101
    102

~/.virtualenvs/incy/lib/python3.8/site-packages/sqlalchemy/util/langhelpers.py in __exit__(self, type_, value, traceback)
     66             self._exc_info = None  # remove potential circular references
     67             if not self.warn_only:
---> 68                 compat.raise_(
     69                     exc_value, with_traceback=exc_tb,
     70                 )

~/.virtualenvs/incy/lib/python3.8/site-packages/sqlalchemy/util/compat.py in raise_(***failed resolving arguments***)
    180
    181         try:
--> 182             raise exception
    183         finally:
    184             # credit to

~/.virtualenvs/incy/lib/python3.8/site-packages/sqlalchemy/orm/loading.py in instances(query, cursor, context)
     86
     87             for path, post_load in context.post_load_paths.items():
---> 88                 post_load.invoke(context, path)
     89
     90             if filtered:

~/.virtualenvs/incy/lib/python3.8/site-packages/sqlalchemy/orm/loading.py in invoke(self, context, path)
    908             ]
    909             if states:
--> 910                 loader(context, path, states, self.load_keys, *arg, **kw)
    911         self.states.clear()
    912

<string> in <lambda>(self, context, path, states, load_only, effective_entity)

~/.virtualenvs/incy/lib/python3.8/site-packages/sqlalchemy/orm/strategies.py in _load_for_path(self, baked, context, path, states, load_only, effective_entity)
   2405
   2406         if query_info.load_only_child:
-> 2407             self._load_via_child(
   2408                 our_states, none_states, query_info, q, context
   2409             )

~/.virtualenvs/incy/lib/python3.8/site-packages/sqlalchemy/orm/strategies.py in _load_via_child(self, our_states, none_states, query_info, q, context)
   2419             chunk = our_keys[0 : self._chunksize]
   2420             our_keys = our_keys[self._chunksize :]
-> 2421             data = {
   2422                 k: v
   2423                 for k, v in q(context.session).params(

~/.virtualenvs/incy/lib/python3.8/site-packages/sqlalchemy/ext/baked.py in __iter__(self)
    443             q = fn(q)
    444
--> 445         return q._execute_and_instances(context)
    446
    447     def count(self):

~/.virtualenvs/incy/lib/python3.8/site-packages/sqlalchemy/orm/query.py in _execute_and_instances(self, querycontext)
   3531         )
   3532
-> 3533         result = conn.execute(querycontext.statement, self._params)
   3534         return loading.instances(querycontext.query, result, querycontext)
   3535

~/.virtualenvs/incy/lib/python3.8/site-packages/sqlalchemy/engine/base.py in execute(self, object_, *multiparams, **params)
   1009             )
   1010         else:
-> 1011             return meth(self, multiparams, params)
   1012
   1013     def _execute_function(self, func, multiparams, params):

~/.virtualenvs/incy/lib/python3.8/site-packages/sqlalchemy/sql/elements.py in _execute_on_connection(self, connection, multiparams, params)
    296     def _execute_on_connection(self, connection, multiparams, params):
    297         if self.supports_execution:
--> 298             return connection._execute_clauseelement(self, multiparams, params)
    299         else:
    300             raise exc.ObjectNotExecutableError(self)

~/.virtualenvs/incy/lib/python3.8/site-packages/sqlalchemy/engine/base.py in _execute_clauseelement(self, elem, multiparams, params)
   1122             )
   1123
-> 1124         ret = self._execute_context(
   1125             dialect,
   1126             dialect.execution_ctx_cls._init_compiled,

~/.virtualenvs/incy/lib/python3.8/site-packages/sqlalchemy/engine/base.py in _execute_context(self, dialect, constructor, statement, parameters, *args)
   1204             context = constructor(dialect, self, conn, *args)
   1205         except BaseException as e:
-> 1206             self._handle_dbapi_exception(
   1207                 e, util.text_type(statement), parameters, None, None
   1208             )

~/.virtualenvs/incy/lib/python3.8/site-packages/sqlalchemy/engine/base.py in _handle_dbapi_exception(self, e, statement, parameters, cursor, context)
   1508                 util.raise_(newraise, with_traceback=exc_info[2], from_=e)
   1509             elif should_wrap:
-> 1510                 util.raise_(
   1511                     sqlalchemy_exception, with_traceback=exc_info[2], from_=e
   1512                 )

~/.virtualenvs/incy/lib/python3.8/site-packages/sqlalchemy/util/compat.py in raise_(***failed resolving arguments***)
    180
    181         try:
--> 182             raise exception
    183         finally:
    184             # credit to

~/.virtualenvs/incy/lib/python3.8/site-packages/sqlalchemy/engine/base.py in _execute_context(self, dialect, constructor, statement, parameters, *args)
   1202                 conn = self._revalidate_connection()
   1203
-> 1204             context = constructor(dialect, self, conn, *args)
   1205         except BaseException as e:
   1206             self._handle_dbapi_exception(

~/.virtualenvs/incy/lib/python3.8/site-packages/sqlalchemy/engine/default.py in _init_compiled(cls, dialect, connection, dbapi_connection, compiled, parameters)
    837                 for key in positiontup:
    838                     if key in processors:
--> 839                         param.append(processors[key](compiled_params[key]))
    840                     else:
    841                         param.append(compiled_params[key])

~/.virtualenvs/incy/lib/python3.8/site-packages/sqlalchemy/dialects/sqlite/base.py in process(value)
    838                 }
    839             else:
--> 840                 raise TypeError(
    841                     "SQLite Date type only accepts Python "
    842                     "date objects as input."

StatementError: (builtins.TypeError) SQLite Date type only accepts Python date objects as input.
[SQL: SELECT user.batch_date AS user_batch_date, user.id AS user_id
FROM user
WHERE (user.batch_date, user.id) IN ([EXPANDING_primary_keys])]
[parameters: [{'primary_keys': [(datetime.date(2020, 10, 18), 1)]}]]
```
for the bindparam query:
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
~/.virtualenvs/incy/lib/python3.8/site-packages/sqlalchemy/engine/base.py in _execute_context(self, dialect, constructor, statement, parameters, *args)
   1203
-> 1204             context = constructor(dialect, self, conn, *args)
   1205         except BaseException as e:

~/.virtualenvs/incy/lib/python3.8/site-packages/sqlalchemy/engine/default.py in _init_compiled(cls, dialect, connection, dbapi_connection, compiled, parameters)
    838                     if key in processors:
--> 839                         param.append(processors[key](compiled_params[key]))
    840                     else:

~/.virtualenvs/incy/lib/python3.8/site-packages/sqlalchemy/dialects/sqlite/base.py in process(value)
    839             else:
--> 840                 raise TypeError(
    841                     "SQLite Date type only accepts Python "

TypeError: SQLite Date type only accepts Python date objects as input.

The above exception was the direct cause of the following exception:

StatementError                            Traceback (most recent call last)
<ipython-input-1-19bad78acc34> in <module>
     62 # It is also similiar construct to the one the selectinload uses.
     63 (
---> 64     session.query(User)
     65     .filter(
     66         tuple_(User.batch_date, User.id)

~/.virtualenvs/incy/lib/python3.8/site-packages/sqlalchemy/orm/query.py in all(self)
   3344                 :ref:`faq_query_deduplicating`
   3345         """
-> 3346         return list(self)
   3347
   3348     @_generative(_no_clauseelement_condition)

~/.virtualenvs/incy/lib/python3.8/site-packages/sqlalchemy/orm/query.py in __iter__(self)
   3506         if self._autoflush and not self._populate_existing:
   3507             self.session._autoflush()
-> 3508         return self._execute_and_instances(context)
   3509
   3510     def __str__(self):

~/.virtualenvs/incy/lib/python3.8/site-packages/sqlalchemy/orm/query.py in _execute_and_instances(self, querycontext)
   3531         )
   3532
-> 3533         result = conn.execute(querycontext.statement, self._params)
   3534         return loading.instances(querycontext.query, result, querycontext)
   3535

~/.virtualenvs/incy/lib/python3.8/site-packages/sqlalchemy/engine/base.py in execute(self, object_, *multiparams, **params)
   1009             )
   1010         else:
-> 1011             return meth(self, multiparams, params)
   1012
   1013     def _execute_function(self, func, multiparams, params):

~/.virtualenvs/incy/lib/python3.8/site-packages/sqlalchemy/sql/elements.py in _execute_on_connection(self, connection, multiparams, params)
    296     def _execute_on_connection(self, connection, multiparams, params):
    297         if self.supports_execution:
--> 298             return connection._execute_clauseelement(self, multiparams, params)
    299         else:
    300             raise exc.ObjectNotExecutableError(self)

~/.virtualenvs/incy/lib/python3.8/site-packages/sqlalchemy/engine/base.py in _execute_clauseelement(self, elem, multiparams, params)
   1122             )
   1123
-> 1124         ret = self._execute_context(
   1125             dialect,
   1126             dialect.execution_ctx_cls._init_compiled,

~/.virtualenvs/incy/lib/python3.8/site-packages/sqlalchemy/engine/base.py in _execute_context(self, dialect, constructor, statement, parameters, *args)
   1204             context = constructor(dialect, self, conn, *args)
   1205         except BaseException as e:
-> 1206             self._handle_dbapi_exception(
   1207                 e, util.text_type(statement), parameters, None, None
   1208             )

~/.virtualenvs/incy/lib/python3.8/site-packages/sqlalchemy/engine/base.py in _handle_dbapi_exception(self, e, statement, parameters, cursor, context)
   1508                 util.raise_(newraise, with_traceback=exc_info[2], from_=e)
   1509             elif should_wrap:
-> 1510                 util.raise_(
   1511                     sqlalchemy_exception, with_traceback=exc_info[2], from_=e
   1512                 )

~/.virtualenvs/incy/lib/python3.8/site-packages/sqlalchemy/util/compat.py in raise_(***failed resolving arguments***)
    180
    181         try:
--> 182             raise exception
    183         finally:
    184             # credit to

~/.virtualenvs/incy/lib/python3.8/site-packages/sqlalchemy/engine/base.py in _execute_context(self, dialect, constructor, statement, parameters, *args)
   1202                 conn = self._revalidate_connection()
   1203
-> 1204             context = constructor(dialect, self, conn, *args)
   1205         except BaseException as e:
   1206             self._handle_dbapi_exception(

~/.virtualenvs/incy/lib/python3.8/site-packages/sqlalchemy/engine/default.py in _init_compiled(cls, dialect, connection, dbapi_connection, compiled, parameters)
    837                 for key in positiontup:
    838                     if key in processors:
--> 839                         param.append(processors[key](compiled_params[key]))
    840                     else:
    841                         param.append(compiled_params[key])

~/.virtualenvs/incy/lib/python3.8/site-packages/sqlalchemy/dialects/sqlite/base.py in process(value)
    838                 }
    839             else:
--> 840                 raise TypeError(
    841                     "SQLite Date type only accepts Python "
    842                     "date objects as input."

StatementError: (builtins.TypeError) SQLite Date type only accepts Python date objects as input.
[SQL: SELECT user.batch_date AS user_batch_date, user.id AS user_id
FROM user
WHERE (user.batch_date, user.id) IN ([EXPANDING_primary_keys])]
[parameters: [{'primary_keys': [(datetime.date(2020, 10, 18), 1)]}]]
```

**Versions.**
 - OS: macOS 11.0 Beta
 - Python: 3.8.6
 - SQLAlchemy: 1.3.19
 - Database: Reproduction with SQLite (I encountered the error originally with PostgreSQL 12.4)
 - DBAPI: pysqlite

**Additional context**
My original problem was with PostgreSQL and composite (UUID, Integer) primary key.
If I change the order of primary keys in the tuple, there is no error. It seems like the bind value processor of the first item in the tuple is used for all items.

**Have a nice day to you too!**

**Comment by zzzeek** [comment]
hi -

this issue has been fixed for 1.4 which is not yet released.   If you confirm by testing against the git master of SQLAlchemy that would be helpful.   issue was fixed as part of #4645  it's mentioned here: https://docs.sqlalchemy.org/en/14/changelog/changelog_14.html#change-ec466d1a10e8ff426e6bd1429f954889    "The change also completes support for expanding IN with tuples, where previously type-specific bind processors weren’t taking effect."

**Comment by tvuotila** [comment]
Thank you for the fast response.

I can confirm that the issue is fixed in the git master.
